A single flow for all outgoing HTTP requests

tareksha opened the pull request:
phase 1 - builder, runner, all json requests

An injected interface, HttpRequestFactory, allows managing all outgoing
HTTP requests. HttpJsonRequest factory should use it internally. The
defult DefaultHttpJsonRequest as reusable as it accepts the general HTTP
factory in its constructor. Extensions can replace HttpJsonRequestFactory
and provide DefaultHttpJsonRequest as-is, given a different impl.

Signed-off-by: Tareq Sharafy tareq.sha@gmail.com
Change-Id: I2f334b7bb133b3f136d2b8e5b18301602f427355

tareksha commented:
@skabashnyuk @evoevodin please review

garagatyi commented on platform-api/che-core-api-runner/src/main/java/org/eclipse/che/api/runner/internal/Runner.java:
```diff
@@ -90,6 +92,10 @@ public boolean isValid(DeploymentSources deployment) {
     private java.io.File             deployDirectory;
 
+    @Inject
+    private DownloadPlugin theDownloadPlugin;
```
Can you explain why we still need old field, looks like it is unused.

tareksha replied:
@garagatyi the old field is final protected, not private. i saw it too risky to just remove it or hide non-incrementally.

garagatyi replied:
I didn't notice that, thanks.

garagatyi commented on platform-api/che-core-api-runner/src/main/java/org/eclipse/che/api/runner/RemoteRunnerProcess.java:
```diff
@@ -128,15 +120,7 @@ public void readRecipeFile(OutputProvider output) throws IOException, RunnerExce
-            conn.setRequestProperty(HttpHeaders.AUTHORIZATION, context.getUser().getToken());
-        }
-        try {
+        try (HttpResponse conn = requestFactory.request(0, url, method, null, null, null)) {
```
I think it is better to keep timeout value to avoid any changes in app behavior

garagatyi commented on platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/rest/HttpRequestFactory.java:
```diff
@@ -0,0 +1,65 @@
+     * @return
+     * @throws IOException
+     */
+    HttpResponse request(int timeout, String url, String method, BodyWriter body, Map<String, String> headers,
```
For me it is very hard to read code with usage of method that has so many parameters. HttpJsonRequestFactory uses another approach with builder like API which is much more readable in several cases.
What do you think about changing general request factory to similar approach?

tareksha replied:
@garagatyi correct, but DefaultHttpJsonRequest uses a non-streaming approach and consumes the body of both requests and responses as full strings, which is not suitable for general HTTP streaming. the change is intended to minimize changes by exposing a common denominator. otherwise the change will be much bigger.

garagatyi replied:
Is it possible to use builder style API of http client with body writer/reader that will implement streaming approach?

garagatyi commented on platform-api/che-core-api-builder/src/main/java/org/eclipse/che/api/builder/RemoteTask.java:
```diff
@@ -217,15 +214,7 @@ public void downloadResultArchive(String archType, HttpOutputMessage output) thr
-            conn.setRequestProperty(HttpHeaders.AUTHORIZATION, context.getUser().getToken());
-        }
-        try {
+        try (HttpResponse conn = requestFactory.request(0, url, null, null, null, null)) {
```
Can you leave old timeout to avoid any changes in app behavior?

garagatyi commented on platform-api/che-core-api-builder/src/main/java/org/eclipse/che/api/builder/RemoteTask.java:
```diff
@@ -217,15 +214,7 @@ public void downloadResultArchive(String archType, HttpOutputMessage output) thr
-        }
-        try {
+        try (HttpResponse conn = requestFactory.request(0, url, null, null, null, null)) {
             output.setStatus(conn.getResponseCode());
```
I don't see where authorization headers is set. Can you point me?

garagatyi commented on platform-api/che-core-api-builder/src/main/java/org/eclipse/che/api/builder/internal/SourcesManagerImpl.java:
```diff
@@ -202,8 +205,6 @@ public void write(byte[] b) throws IOException {
     private void download(String downloadUrl, java.io.File downloadTo, File directory) throws IOException {
-        HttpURLConnection conn = null;
-        try {
             final LinkedList<java.io.File> q = new LinkedList<>();
```
Since try is removed looks like indentation is too big

tareksha replied:
@garagatyi I did this to avoid bumping the whole file to the left and affect only the specific lines that changed. I can un-indent it.

garagatyi replied:
I think github will show that changes are in indentation only

garagatyi commented on platform-api/che-core-api-core/src/main/java/org/eclipse/che/api/core/rest/DefaultHttpJsonRequest.java:
```diff
@@ -181,48 +182,24 @@ protected DefaultHttpJsonResponse doRequest(int timeout,
                                                                                BadRequestException {
-        final String authToken = getAuthenticationToken();
-        final boolean hasQueryParams = parameters != null && !parameters.isEmpty();
-        if (hasQueryParams || authToken != null) {
```
Since authentication is dropped this change will change app behavior. Looks like it is not very safe for stability of the project

garagatyi replied:
Sorry, now I see that you've moved this code.

tareksha commented:
@garagatyi  the change now uses the same builder pattern. please review

tareksha committed:
A single flow for all outgoing HTTP requests

phase 1 - builder, runner, all json requests

A new builder inteface HttpRequest, sharing features with the existing
HttpJsonRequest, allows building general requests and consuming responses
through HttpResponse. A single HTTP flow is achieving by sharing a single
concrete HTTP request execution method.

Signed-off-by: Tareq Sharafy <tareq.sha@gmail.com>
Change-Id: I2f334b7bb133b3f136d2b8e5b18301602f427355

tareqhs commented:
@garagatyi @skabashnyuk @vparfonov @evoevodin it passes all tests

skabashnyuk commented:
@tareqhs Quick response is -1 to make HttpJsonRequest something not related to json. We understand that you want to make some tooling to make general request but making it in HttpJsonRequest is not that we want to achieve. I hope @garagatyi @evoevodin  can add more to this statement.

voievodin commented:
Hi @tareqhs.
Generally speaking `HttpJsonRequest` was designed to work with json only, there are several reasons for that
- 90% requests are requests between our API services where json is a main format, so it fits almost all the needs
- when the tool is json specialized we can add useful methods related to DTO processing, which is also json based, so the tool becomes even more convenient and usable.

The idea of you changes is clear for me and makes total sense, but the implementation strategy should be different, as `HttpJsonRequest` becomes not really Json and the client of the class may be confused as the class contains general and json request logic mixed.

As for me the best solution here is to create  general class like `HttpRequest` with all the general methods for requesting stuff through http and then change the current `HttpJsonRequest` implementation to use `HttpRequest` in this way the logic is clearly separated between components.

tareksha commented:
hi @evoevodin @skabashnyuk ,

Well, that was the original approach in the PR but I tried to minimize code changes. The main issue is not whether to extend `HttpJsonRequest` itself or not, but rather the factory `HttpJsonRequestFactory`. I'm seeking a solution that doesn't require injecting a new factory everywhere yet again, which conceptually is a superset of `HttpJsonRequestFactory`.

This is my suggestion : `HttpJsonRequestFactory` extends a general `HttpRequestFactory` that has general request methods, so that the existing injections stay identical except for receiving the additional HTTP APIs that are in turn used instead of `HttpURLConnection` where needed.

Agree?
